packages/amplify-graphql-conversation-transformer/src/utils/tools.ts

```diff
@@ -0,0 +1,244 @@
+  seenTypes.add(type.name.value);
+
+  // If the type has no fields, return an empty string
+  if (!type.fields || type.fields.length === 0) {
+    return '';
+  }
+
+  // Build the selection set for this type
+  const fieldSelections = type.fields
+    .map((field) => {
+      if (isScalar(field.type)) {
+        // For scalar fields, just include the field name
+        return field.name.value;
+      } else {
+        // For object types, recursively generate the nested selection
+        const nestedSelection = generateSelectionSet(field.type, ctx, new Set(seenTypes), field.name.value);
+        return nestedSelection ? `${field.name.value} { ${nestedSelection} }` : '';
+      }
+    })
+    .filter(Boolean); // Remove any empty strings
+
+  return fieldSelections.join(' ');
+};
+
+/**
+ * Retrieves the ObjectTypeDefinitionNode for a given type name from the context.
+ *
+ * @param {string} name - The name of the type to retrieve.
+ * @param {TransformerContextProvider} ctx - The transformer context provider.
+ * @returns {ObjectTypeDefinitionNode} The object type definition node.
+ * @throws {Error} If the type definition cannot be found.
+ */
+const getObjectTypeFromName = (name: string, ctx: TransformerContextProvider): ObjectTypeDefinitionNode => {
+  const node = ctx.output.getObject(name);
+  if (!node) {
+    throw new Error(`Could not find type definition for ${name}`);
+  }
+  return node;
+};
+
+/**
+ * Processes tool definitions and generates a Tools object.
+ *
+ * @param {ToolDefinition[]} toolDefinitions - An array of tool definitions.
+ * @param {TransformerContextProvider} ctx - The transformer context provider.
+ * @returns {Tools | undefined} A Tools object if valid tool definitions are provided, undefined otherwise.
+ * @throws {InvalidDirectiveError} If there are no queries or if a tool is defined without a matching Query field.
+ */
+export const processTools = (toolDefinitions: ToolDefinition[], ctx: TransformerContextProvider): Tools | undefined => {
+  // Early return if no tool definitions are provided
+  if (!toolDefinitions || toolDefinitions.length === 0) {
+    return undefined;
+  }
+
+  // Retrieve Query type fields
+  const queryType = ctx.output.getType('Query') as ObjectTypeDefinitionNode;
+  if (!queryType.fields || queryType.fields.length === 0) {
+    throw new InvalidDirectiveError('Tools must be queries - no queries found in the schema');
+  }
+
+  // Process each tool definition
+  const tools: Tool[] = toolDefinitions.map((toolDefinition) => {
+    const { name: toolName, description } = toolDefinition;
+    const queryField = queryType.fields?.find((field) => field.name.value === toolName);
+
+    if (!queryField) {
+      throw new InvalidDirectiveError(`Tool "${toolName}" defined in @conversation directive has no matching Query field definition`);
+    }
+
+    return createTool(toolName, description, queryField, ctx);
+  });
+
+  return { tools };
```

Comment:
I initially added the types because they represent a part of the lambda event payload we're sending from the resolver, so the types give us the opportunity to ensure correctness here. I get how it's confusing as is though. I'll see about improving this in a follow up. Thanks for the callout.